core/src/main/clojure/xtdb/object_store.clj

```diff
@@ -61,102 +59,56 @@
-            new-pos (+ (.position buf) (int start))]
-        (.slice buf new-pos (int (max 1 (min (- (.remaining buf) new-pos) len)))))))
-
-  (putObject [_this k buf]
-    (.putIfAbsent os k (.slice buf))
-    (CompletableFuture/completedFuture nil))
-
-  (listObjects [_this]
-    (vec (.keySet os)))
+    (util/->mmap-path from-path)))
 
-  (listObjects [_this prefix]
-    (->> (.keySet (.tailMap os prefix))
-         (into [] (take-while #(str/starts-with? % prefix)))))
+(defn get-path-range ^java.nio.ByteBuffer [^Path root-path, ^String k, ^long start, ^long len]
+  (ensure-shared-range-oob-behaviour start len)
 
-  (deleteObject [_this k]
-    (.remove os k)
-    (CompletableFuture/completedFuture nil))
+  (let [from-path (.resolve root-path k)]
+    (when-not (util/path-exists from-path)
+      (throw (obj-missing-exception k)))
 
-  Closeable
-  (close [_]
-    (.clear os)))
-
-(defmethod ig/init-key ::memory-object-store [_ _]
-  (->InMemoryObjectStore (ConcurrentSkipListMap.)))
+    (with-open [in (util/->file-channel from-path #{:read})]
+      (.map in FileChannel$MapMode/READ_ONLY start (max 1 (min (- (.size in) start) len))))))
 
-(defmethod ig/halt-key! ::memory-object-store [_ ^InMemoryObjectStore os]
-  (.close os))
+(defn put-path [^Path root-path, ^String k, ^ByteBuffer buf]
+  (let [buf (.duplicate buf)
+        to-path (.resolve root-path k)]
+    (util/mkdirs (.getParent to-path))
 
-(derive ::memory-object-store :xtdb/object-store)
+    (if (identical? (FileSystems/getDefault) (.getFileSystem to-path))
+      (if (util/path-exists to-path)
+        to-path
+        (util/write-buffer-to-path-atomically buf root-path to-path))
 
-(comment
+      (util/write-buffer-to-path buf to-path))))
 
-  (def mos
-    (->> (ig/prep-key ::memory-object-store {})
-         (ig/init-key ::memory-object-store)))
+(defn list-path
+  ([^Path root-path]
+   (with-open [dir-stream (Files/walk root-path (make-array FileVisitOption 0))]
+     (vec (sort (for [^Path path (iterator-seq (.iterator dir-stream))
+                      :when (Files/isRegularFile path (make-array LinkOption 0))]
+                  (str (.relativize root-path path)))))))
 
-  (.close mos)
-
-  @(.putObject mos "foo.txt" (ByteBuffer/wrap (.getBytes "hello, world!")))
-
-  (let [buf @(.getObject mos "foo.txt")
-        arr (byte-array (.remaining buf))]
-    (.get buf arr)
-    (String. arr))
-
-  (let [buf @(.getObjectRange mos "foo.txt" 2 5)
-        arr (byte-array (.remaining buf))]
-    (.get buf arr)
-    (String. arr))
-
-  )
+  ([^Path root-path, ^String dir]
+   (let [dir (.resolve root-path dir)]
+     (when (Files/exists dir (make-array LinkOption 0))
+       (with-open [dir-stream (Files/newDirectoryStream dir)]
+         (vec (sort (for [^Path path dir-stream]
+                      (str (.relativize root-path path))))))))))
 
 (deftype FileSystemObjectStore [^Path root-path, ^ExecutorService pool]
```

Comment:
Presumably this is now our remote filesystem store implementation (thus why we're using all the calls to `Files` and whatnot)?.